RandomStateAntipattern.md

```diff
@@ -0,0 +1,383 @@
+as in this buggy example (the third call should have been `nextLong(ss)`).
+
+Even worse, since we're using random numbers where the number stream and the seed
+have the same type, without extra machinery you can even get the order of arguments wrong:
+
+```scala
+val sum3pof_verywrong = (l: Long) => {
+  val (a, s) = nextLong(l)
+  val (b, ss) = nextLong(s)
+  val (c, sss) = nextLong(ss)
+  (a + b + c, sss)
+}
+```
+
+Now we've mixed up our seed and our computation, and we generated and added the
+same number three times all because the local context did not make it apparent
+which of the two results was which.  (You can avoid this by wrapping the seed
+to make it unambiguously different than the random number return type.  But
+there's no point, because the exposed states are _still_ error-prone in the
+first way; we just shouldn't do this at all.)
+
+For random numbers, where independence is (usually) everything, this is as bad
+as it gets.  Avoid this pattern like the plague!  If you must have referential
+transparency in your random numbers, use `State` or some other mechanism to
+enforce the discipline required to keep your state safely progressing!
+
+## Near-parity again: accumulation
+
+Let's suppose we want to generate more than just `Long` numbers.  Maybe we want to generate a string
+containing a random sequence of `t` and `f`.  Let's give it a go each way.
+
+First, for our mutable RNG, since we didn't bother to give `nextBool` and `nextString(length: Int)` methods, let's add them.
+
+We have a `fill` method that lets us repeat an operation a certain number of times, so we can:
+
+```scala
+implicit class MoreRandomStuff(private val r: Rng) extends AnyVal {
+  def nextBool = r.nextLong < 0
+  def nextTF = if (nextBool) 't' else 'f'
+  def nextString(length: Int) = new String(Array.fill(length)(nextTF))
+}
+```
+
+Not hard at all!
+
+Let's try with the monadic version.  We can't just use a `for` comprehension, because it doesn't support
+generating arbitrarily nested calls to `flatMap`.  Instead, we can turn to a method called `replicateA` that
+does something similar (the `A` part meaning that you're replicating just the output, not the seeds)
+
+The trick with `sequence` is that it will turn `F[G[A]]` into `G[F[A]]`, that is swap the wrapper order, assuming that
+`G` and `F` are compatible.  It turns out that `State` confuses type inference, but we can explicitly ask for what we
+want, and then it works.  Also, because we're swapping two monads, we need to build up all the operations we want inside
+the same kind of collection we want to end up with.
+
+
+```scala
+import cats.implicits._
+val nextBool = nextLong.map(_ < 0)
+val nextTF = nextBool.map(b => if (b) 't' else 'f')
+def nextString(length: Int) = nextTF.replicateA(length).map(_.mkString)
+```
+
+This wasn't terribly arduous, but it also wasn't free.  We needed an extra concept (`replicateA`--just because we
+have `replicateA` doesn't mean we can forget about `fill`!).
+
```

Comment:
Part of my point is that you have more concepts to work with simultaneously mentally, which is both taxing and inelegant.  Only a _little_ bit taxing--we're pretty good with dealing with lots of concepts--and plenty of people prefer a large collection of special-purpose tools rather than a single elegant solution.  But it's weirdly backwards from the normal meme of FP requiring more general/abstract thinking (and I generally agree that's not a fair criticism if you can _restrict_ your thinking to the abstract level and don't have to deal with the mucky details at all).  Here, it's the opposite!  You have _more_ futzing with mucky details because of the constraints embodied by `State`.

Reply:
I just don't think you're counting fairly.  You have to know whether you are dealing with applicative actions or not to know whether to use `replicateA`.  If you get it wrong, the compiler will stop you, but it won't fix it for you.  You still have to think it through.

In contrast, with mutable Rng, all the values being produced are just values, and work just like any other values.  They happen to be picked randomly, but that's irrelevant.

I suppose in a sense there is conservation of number of concepts; you get to not think about state, but the person thinking about state doesn't have to think about the type of the computation in addition to the type of what is being computed.